src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseInviteService.cs

```diff
@@ -0,0 +1,227 @@
+                .Select(urr => urr.ReleaseId)
+                .ToList();
+
+            var missingReleaseRoleReleaseIds = releaseIds
+                .Where(releaseId => !existingReleaseRoleReleaseIds.Contains(releaseId))
+                .ToList();
+
+            if (missingReleaseRoleReleaseIds.IsNullOrEmpty())
+            {
+                return ValidationActionResult(UserAlreadyHasReleaseRoles);
+            }
+
+            var emailResult = await SendContributorInviteEmail(
+                publicationTitle: publicationTitle,
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email);
+            if (emailResult.IsLeft)
+            {
+                return emailResult;
+            }
+
+            await _userReleaseRoleRepository.CreateManyIfNotExists(
+                userId: userId,
+                releaseIds: missingReleaseRoleReleaseIds,
+                role: Contributor,
+                createdById: _userService.GetUserId()
+            );
+
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId(),
+                accepted: true);
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Publication>> ValidateReleaseIds(Publication publication,
+            List<Guid> releaseIds)
+        {
+            var distinctReleaseIds = releaseIds.Distinct().ToList();
+            if (distinctReleaseIds.Count != releaseIds.Count)
+            {
+                throw new ArgumentException($"{nameof(releaseIds)} should not contain duplicates",
+                    nameof(releaseIds));
+            }
+
+            await _contentDbContext
+                .Entry(publication)
+                .Collection(p => p.Releases)
+                .LoadAsync();
+
+            var publicationReleaseIds = publication
+                .ListActiveReleases()
+                .Select(r => r.Id)
+                .ToList();
+
+            if (!releaseIds.All(publicationReleaseIds.Contains))
+            {
+                return ValidationActionResult(NotAllReleasesBelongToPublication);
+            }
+
+            return publication;
+        }
+
+        private async Task<Either<ActionResult, Unit>> SendContributorInviteEmail(
+            string publicationTitle, List<Guid> releaseIds, string email)
+        {
+            if (releaseIds.IsNullOrEmpty())
+            {
+                throw new ArgumentException("List of releases cannot be empty");
+            }
+
+            var template = _configuration.GetValue<string>("NotifyContributorTemplateId");
+            var scheme = _httpContextAccessor.HttpContext.Request.Scheme;
+            var host = _httpContextAccessor.HttpContext.Request.Host;
+            var url = $"{scheme}://{host}/";
```

Comment:
This doesn't look right - we set an `AdminUri` variable in configuration that should be used instead. See `EmailTemplateService` for examples where we're doing just that.

Using the context accessor could be a problem as the URL will be controlled by the request itself. It's possible (although unlikely) that a user can directly access the API through a non-DNS URL and can consequently generate an incorrect URL.